pkg/ebpf/events/derived/derived.go

```diff
@@ -0,0 +1,29 @@
+package derived
+
+import "github.com/aquasecurity/tracee/types/trace"
+
+// DeriveFn is a function prototype for a function that receives an event as
+// argument and may produce a new event if relevant.
+// It returns list of derived events, which might be empty depending on successful derivation,
+// a bool indicating if an event was derived, and an error if one occurred.
+type DeriveFn func(trace.Event) ([]trace.Event, bool, error)
+
+// IDerivedEventGenerator is an interface for derived events generator class.
+// The main purpose of it is to create derived events given an event.
+// To generate derived event, the GenerateDerivedFn function should be used with initialized generator.
+type IDerivedEventGenerator interface {
```

Comment:
Do we use this anywhere besides GenerateDerivedFn? If not, where would it be used?

Reply:
No, I didn't plan for it to be used anywhere except for `GenerateDerivedFn`.
But this is the idea, I want 1 structure for creating derivation function to make the code using it more readable.

However, I think we can also make the `eventsDerivation` map use the interface instead of the function.
This will make it possible to have state of single derived event generator that use multiple base events for its derivation.
So I propose using a map of the following structure - `map[int32]map[int32]IDerivedEventGenerator`.
Afterwards we will call the `GenerateDerivedFn` function for each struct and create the function with state between different triggering event. WDYT?

Reply:
Well, that's the purpose of the DeriveFn, to be the interface. 
If you need some external dependency, we have an existing pattern to support that (we use that with tracee for example for the container events).

It could be that what you suggest would be more readable, but we can't actually see that if we don't write it out. So if you could maybe rewrite one of the derivation functions to use the IDerivedEventGenerator, that may give an idea if this would make it actually more readable.